src/Features/CSharp/Portable/CodeMapper/CSharpMapCodeService.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeCleanup;
+using Microsoft.CodeAnalysis.CodeMapping;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeMapping;
+
+/// <summary>
+/// A mapper used for Mapping C# Code from a suggestion onto existing code in the codebase.
+/// This is the backbone of previews, as this allows us to know where we should place the preview.
+/// </summary>
+[ExportLanguageService(typeof(IMapCodeService), language: LanguageNames.CSharp), Shared]
+internal sealed partial class CSharpMapCodeService : IMapCodeService
+{
+    private readonly IGlobalOptionService _globalOptions;
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public CSharpMapCodeService(IGlobalOptionService globalOptions)
+    {
+        _globalOptions = globalOptions;
+    }
+
+    public async Task<Document> MapCodeAsync(Document document, ImmutableArray<string> contents, ImmutableArray<DocumentSpan> focusLocations, bool formatMappedCode, CancellationToken cancellationToken)
+    {
+        var target = await GetMappingTargetAsync(document, focusLocations, cancellationToken).ConfigureAwait(false);
+
+        if ((await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false))?.Options is not CSharpParseOptions options)
+            return document;
+
+        using var _ = ArrayBuilder<TextChange>.GetInstance(out var changes);
+        foreach (var code in contents)
+        {
+            var mapHelper = await GetMapperHelperAsync(target, code, options, cancellationToken).ConfigureAwait(false);
+            if (mapHelper is not null)
+            {
+                var edits = await mapHelper.MapCodeAsync(cancellationToken).ConfigureAwait(false);
+                changes.AddRange(edits);
+            }
+        }
+
+        return await GetDocumentWithChangeAsync(document, changes.ToImmutable(), formatMappedCode, cancellationToken).ConfigureAwait(false);
+    }
+
+    private async Task<Document> GetDocumentWithChangeAsync(Document document, ImmutableArray<TextChange> textChanges, bool formatMappedCode, CancellationToken cancellationToken)
```

Comment:
Is this to re-assemble the current session document with the changes provided by the mapper call?
I wonder if we could abstract this so that it can be used by other code mappers too, since all of them will require to do something similar.

Reply:
> Is this to re-assemble the current session document with the changes provided by the mapper call

yea. Since `IMapCodeService` is a Roslyn language service, the API of follows the convention that accept a Document and returns tranformed Document. When this service consumed by VS Copilot chat, it will need to get the TextChanges from this new doc (i.e. that's what the C# LSP handler does in this PR)

> I wonder if we could abstract this so that it can be used by other code mappers too

Sure, we are using Roslyn types here, but I guess the core logic is sharable if needed.